File: Data Structures & Algorithms/invert-a-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        headKeeper = root
        self.invertAux(root)
        return headKeeper

    def invertAux(self, root : Optional[TreeNode]) -> Optional[TreeNode]:
        if root is None:
            return None

        tmpLeftNode = root.left
        root.left = root.right
        root.right = tmpLeftNode

        self.invertAux(root.left)
        self.invertAux(root.right)

        return root

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) {\ this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val; 
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        if(root == null) return 0;

        Stack<TreeNode> stack = new Stack<>();
        Stack<Integer> maxStack = new Stack<>();
        stack.push(root);
        maxStack.push(root.val);
        int nGoodNodes = 0;

        while(!stack.isEmpty()){
            TreeNode currNode = stack.pop();
            int currMax = maxStack.pop();
            
            if(currNode.val >= currMax){
                nGoodNodes++;
            }
            
            int nextMax = Math.max(currMax, currNode.val);

            if(currNode.right != null) {
                stack.push(currNode.right);
                maxStack.push(nextMax);
            }
            if(currNode.left != null) {
                stack.push(currNode.left);
                maxStack.push(nextMax);
            }
        }
        return nGoodNodes;
    }
}